pkg/test/impl/driver/instance.go

```diff
@@ -177,32 +159,39 @@ func (d *driver) Run() int {
 
 	d.running = false
 
-	d.doCleanup()
-	return rt
-}
+	d.ctx.Tracker().Cleanup()
 
-// GetContext returns the internal test context.
-func (d *driver) GetContext() internal.TestContext {
-	return d
+	return rt
 }
 
-// Environment implements same-named Interface method.
-func (d *driver) Environment() environment.Interface {
-	scope.Debugf("Enter: driver.Environment (%s)", d.testID)
+func (d *driver) AcquireEnvironment(t testing.TB) environment.Interface {
+	t.Helper()
+	scope.Debugf("Enter: driver.AcquireEnvionment (%s)", d.ctx.TestID())
 	d.lock.Lock()
 	defer d.lock.Unlock()
 
-	if d.running {
-		return d.env
+	if !d.running {
+		t.Fatalf("Test driver is not running.")
 	}
 
-	return nil
+	// Double check the test name to see if this is a singleton call for this test?
+	if _, ok := d.testNames[t.Name()]; ok {
+		t.Fatalf("AcquireEnvironment should be called only once during a test session. (test='%s')", t.Name())
+	}
+	d.testNames[t.Name()] = struct{}{}
+
+	// Reset all resettables, as we're going to be executing within the context of a new test.
+	if err := d.ctx.Tracker().Reset(d.ctx); err != nil {
+		t.Fatalf("AcquireEnvironment failed to reset the resource state: %v", err)
+	}
+
+	return d.ctx.Environment()
 }
 
 // InitializeTestDependencies implements same-named Interface method.
-func (d *driver) InitializeTestDependencies(t testing.TB, dependencies []dependency.Dependency) {
+func (d *driver) InitializeTestDependencies(t testing.TB, dependencies []dependency.Instance) {
```

Comment:
nit: Not sure exactly how Go handles ellipsis vs arrays.  I know you can pass an array into an argument that uses ellipsis, but not sure the other way around. Can you do the following with an array parameter?

```golang
e.InitializeTestDependencies(t, d1, d2, d3)
```

Reply:
Internally, the variadic function parameter should be getting treated as an array. Otherwise the type-checker would barf.
https://play.golang.org/p/jRj-xTSQzqt

Reply:
Right, if the method is defined with varargs it will convert to an array.  However, if you define the parameter as an array, you can't pass in varargs.  Just for simplicity of the UI, it might be nice to define the method as:

```golang
InitializeTestDependencies(t testint.TB, dependencies ...int)
```

That way we can call it either way:

```golang
e.InitializeTestDependencies(t, d1, d2, d3)
e.InitializeTestDependencies(t, []dependency.Instance{d1, d2, d3})
```

Reply:
This is the internal API of the driver. It shouldn't be getting called directly. The public API already uses ellipsis:
https://github.com/istio/istio/blob/4548dc7f74900be23c3e21d2bea6f99bb9f502e6/pkg/test/operations.go#L73